src/hw1/bash.py

```diff
@@ -0,0 +1,39 @@
+def head(filename="first.txt", lines=10):
+    if line < 0:
+        raise RuntimeError("Number of lines must be positive")
+    with open(filename, "rt") as file:
+        file_head = [file.readline() for i in range(lines)]
+    print("".join(file_head))
+
+
+def tail(filename="first.txt", lines=10):
+    if line < 0:
+        raise RuntimeError("Number of lines must be positive")
+    with open(filename, "rt") as file:
+        file_backward = [line for line in reversed(file.readlines())]
+        file_backward[0] += "\n"
+        file_backward[-1] = file_backward[-1][0:-1]
+        print("".join(file_backward[:lines]))
+
+
+def nl(filename="first.txt"):
+    with open(filename, "rt") as file:
+        lines = [line[0:-1] for line in file.readlines()]
```

Comment:
This is not memory-efficient, because you are reading the entire file. Instead, you can just iterate with `for line in file`.